modules/index/src/main/java/org/apache/ignite/internal/idx/InternalSortedIndexImpl.java

```diff
@@ -0,0 +1,270 @@
+ * limitations under the License.
+ */
+
+package org.apache.ignite.internal.idx;
+
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.BitSet;
+import java.util.Iterator;
+import java.util.List;
+import java.util.Set;
+import java.util.function.Function;
+import java.util.stream.Collectors;
+import org.apache.ignite.internal.schema.BinaryRow;
+import org.apache.ignite.internal.schema.Column;
+import org.apache.ignite.internal.storage.index.IndexBinaryRow;
+import org.apache.ignite.internal.storage.index.IndexRow;
+import org.apache.ignite.internal.storage.index.SortedIndexColumnDescriptor;
+import org.apache.ignite.internal.storage.index.SortedIndexStorage;
+import org.apache.ignite.internal.table.StorageRowListener;
+import org.apache.ignite.internal.table.TableImpl;
+import org.apache.ignite.internal.table.TableRow;
+import org.apache.ignite.internal.util.Cursor;
+import org.apache.ignite.lang.IgniteInternalException;
+import org.apache.ignite.lang.IgniteUuid;
+import org.apache.ignite.table.Tuple;
+import org.jetbrains.annotations.NotNull;
+import org.jetbrains.annotations.Nullable;
+
+/**
+ * Internal index manager facade provides low-level methods for indexes operations.
+ */
+public class InternalSortedIndexImpl implements InternalSortedIndex, StorageRowListener {
+    private final IgniteUuid id;
+
+    private final String name;
+
+    private final TableImpl tbl;
+
+    private final SortedIndexStorage store;
+
+    /**
+     * Create sorted index.
+     */
+    public InternalSortedIndexImpl(IgniteUuid id, String name, SortedIndexStorage store, TableImpl tbl) {
+        this.id = id;
+        this.name = name;
+        this.store = store;
+        this.tbl = tbl;
+    }
+
+    /** {@inheritDoc} */
+    @Override
+    public IgniteUuid id() {
+        return id;
+    }
+
+    /** {@inheritDoc} */
+    @Override
+    public String name() {
+        return name;
+    }
+
+    /** {@inheritDoc} */
+    @Override
+    public String tableName() {
+        return tbl.name();
+    }
+
+    /** {@inheritDoc} */
+    @Override
+    public List<Column> columns() {
+        return store.indexDescriptor().columns().stream()
+                .map(SortedIndexColumnDescriptor::column)
+                .collect(Collectors.toList());
+    }
+
+    /** {@inheritDoc} */
+    @Override
+    public Cursor<Tuple> scan(Tuple low, Tuple up, byte scanBoundMask, BitSet proj) {
+        Cursor<IndexRow> cur = store.range(
+                low != null ? new IndexSearchRow(low) : null,
+                up != null ? new IndexSearchRow(up) : null,
+                r -> true
+        );
+
+        List<String> tblColIds = new ArrayList<>(tbl.schemaView().schema().columnNames());
+        Set<String> idxColIds = columns().stream().map(Column::name).collect(Collectors.toSet());
+
+        boolean needLookup = proj.stream().anyMatch(order -> !idxColIds.contains(tblColIds.get(order)));
```

Comment:
```suggestion
        boolean indexOnly = proj.stream().map(tblColIds::get).allMatch(idxColIds::contains);
```
Or better avoid streams.